File: AudioKit/Operations/Signal Generators/Segment Arrays/AKSegmentArray.m
//
//  AKSegmentArray.m
//  AudioKit
//
//  Created by Aurelius Prochazka on 1/4/15.
//  Copyright (c) 2015 Aurelius Prochazka. All rights reserved.
//

#import "AKSegmentArray.h"

@implementation AKSegmentArray
{
    AKConstant *value1;
    NSMutableArray<AKConstant *> *segments;
    NSArray<AKConstant *> *releaseSegment;
}

- (instancetype)initWithInitialValue:(AKConstant *)initialValue
                         targetValue:(AKConstant *)targetValue
                       afterDuration:(AKConstant *)duration
                           concavity:(AKConstant *)concavity

{
    self = [super initWithString:[self operationName]];
    if (self) {
        value1      = initialValue;
        segments = [[NSMutableArray alloc] init];
        releaseSegment = [NSArray array];
        [self addValue:targetValue afterDuration:duration concavity:concavity];
        self.state = @"connectable";
    }
    
    return self;
}

- (void)addValue:(AKConstant *)value
   afterDuration:(AKConstant *)duration
       concavity:(AKConstant *)concavity
{
    [segments addObject:duration];
    [segments addObject:concavity];
    [segments addObject:value];
}

- (instancetype)initWithInitialValue:(AKConstant *)initialValue
                         targetValue:(AKConstant *)targetValue
                       afterDuration:(AKConstant *)duration
{
    return [self initWithInitialValue:initialValue
                          targetValue:targetValue
                        afterDuration:duration
                            concavity:akp(1)];
}

- (void)addValue:(AKConstant *)value afterDuration:(AKConstant *)duration
{
    [self addValue:value afterDuration:duration concavity:akp(1)];
}

- (void)releaseToValue:(AKConstant *)value
         afterDuration:(AKConstant *)duration
             concavity:(AKConstant *)concavity
{
    releaseSegment = [NSArray arrayWithObjects:duration, concavity, value, nil];
}

- (NSString *)stringForCSD {
    NSString *opcode = @"transeg";
    if ([releaseSegment count] > 0) {
        opcode = @"transegr";
        [segments addObjectsFromArray:@[releaseSegment[0], releaseSegment[1], releaseSegment[2]]];
    }
    return [NSString stringWithFormat:
            @"%@ %@ %@, %@",
            self,
            opcode,
            value1,
            [segments componentsJoinedByString:@", "]];
}

@end

File: Tests/Tests/AKSegmentArray.swift
//
//  main.swift
//  AudioKit
//
//  Created by Aurelius Prochazka on 1/4/15.
//  Copyright (c) 2015 Aurelius Prochazka. All rights reserved.
//

import Foundation

let testDuration: NSTimeInterval = 10

class Instrument : AKInstrument {

    override init() {
        super.init()

        let segments = AKSegmentArray(
            initialValue: 440.ak,
            targetValue: 660.ak,
            afterDuration: 2.ak,
            concavity: 3.ak
        )
        segments.addValue(550.ak, afterDuration: 0.ak, concavity:0.ak)
        segments.addValue(550.ak, afterDuration: 1.ak, concavity:0.ak)
        segments.addValue(880.ak, afterDuration: 0.ak, concavity:0.ak)
        segments.addValue(220.ak, afterDuration: 6.ak, concavity:(-5).ak)
        segments.addValue(220.ak, afterDuration: 1.ak, concavity:0.ak)

        enableParameterLog("segment value = ", parameter: segments, timeInterval: 0.1)

        let oscillator = AKOscillator()
        oscillator.frequency = segments
        setAudioOutput(oscillator)
    }
}

AKOrchestra.testForDuration(testDuration)

let instrument = Instrument()
AKOrchestra.addInstrument(instrument)

instrument.play()

NSThread.sleepForTimeInterval(NSTimeInterval(testDuration))
